packages/collabswarm/src/collabswarm-document.ts

```diff
@@ -800,47 +812,45 @@ export class CollabswarmDocument<
     // Apply local change w/ automerge.
     this._document = newDocument;
 
-    // Store changes in ipfs.
-    const hash = await this._putBlock(changes);
-    this._hashes.add(hash);
+    await this._makeChange(changes);
+  }
 
-    // Send new message.
-    let updateMessage = this._createSyncMessage();
-    const changeNode: CRDTChangeNode<ChangesType> = {
-      change: changes,
-    };
-    if (updateMessage.changeId && updateMessage.changes) {
-      // TODO: Add links to other part of change tree (See Merkle CRDT paper section VI.B.e).
-      changeNode.children = {};
-      changeNode.children[updateMessage.changeId] = updateMessage.changes;
+  public async addWriter(writer: PublicKey) {
+    // Check that we are a writer (allowed to write to this document).
+    if (!(await this._writers.check(this._userPublicKey))) {
+      throw new Error(
+        `Current user does not have write permissions for: ${this.documentPath}`,
+      );
     }
-    updateMessage.changeId = hash;
-    updateMessage.changes = changeNode;
 
-    // Sign new message.
-    updateMessage.signature = await this._signAsWriter(updateMessage);
+    // Check that the writer is not already a writer.
+    if (this._writers.check(writer)) {
+      return;
+    }
 
-    this._lastSyncMessage = updateMessage;
-    const serializedUpdate = this._syncMessageSerializer.serializeSyncMessage(
-      updateMessage,
-    );
+    // Construct a new writer ACL change.
+    const changes = await this._writers.add(writer);
 
-    // Encrypt sync message.
-    const documentKey = await this._keychain.current();
-    if (!documentKey) {
-      throw new Error(`Document ${this.documentPath} has an empty keychain!`);
+    await this._makeChange(changes);
+  }
+
+  public async removeWriter(writer: PublicKey) {
+    // Check that we are a writer (allowed to write to this document).
+    if (!(await this._writers.check(this._userPublicKey))) {
+      throw new Error(
+        `Current user does not have write permissions for: ${this.documentPath}`,
+      );
     }
-    const { nonce, data } = await this._authProvider.encrypt(
-      serializedUpdate,
-      documentKey,
-    );
-    await this.swarm.ipfsNode.pubsub.publish(
-      this.documentPath,
-      nonce ? concatUint8Arrays(nonce, data) : data,
-    );
 
-    // Fire change handlers.
-    this._fireLocalUpdateHandlers([hash]);
+    // Check that the writer is already a writer.
+    if (!this._writers.check(writer)) {
+      return;
+    }
+
+    // Construct a new writer ACL change.
+    const changes = await this._writers.add(writer);
```

Comment:
`const changes = await this._writers.remove(writer);`

Reply:
```suggestion
    const changes = await this._writers.remove(writer);
```